### ISSUE TITLE
Folders + Filenames truncated when they contain SharPoint site name

### ISSUE THREAD

**Issue by sg-chrishasz** [koltyakov/sppull#27]
I'm seeing a case in SPGo, where the [Retrieve Folder](https://docs.sitego.co/spgo/commands/retrieve-folder) and [Populate Workspace](https://docs.sitego.co/spgo/commands/populate-workspace) commands are incorrectly renaming or repathing files when a that file or folder contains the subsite name. This issue only occurs when using SPPull against a subsite.

For example, if you try to download the entire contents of the `_catalogs/masterpage/display templates/` folder from the `search` root site collection subsite, all files from the `_catalogs/masterpage/display templates/search/` folder will be downloaded to the `_catalogs/masterpage/display templates/` folder (omitting the `/search` path

alternatively, if the filename contains the subsite name as a substring of it's filename, then that token will be removed from the downloaded file ( from: https://github.com/readysitego/spgo/issues/69 )

> The real "Site Assets" structure on the server is:
> SiteAssets/js/**subsite**Name111.xxx.js
> SiteAssets/js/**subsite**Name222.yyy.js
> 
> File names start with the same name as the subsiteName.
> 
> After populating the workspace the js files are shown outside the js folder and are renamed. The populated structure is the following:
> SiteAssets/js111.xxx.js
> SiteAssets/js222.yyy.js

 

See related issues:
https://github.com/readysitego/spgo/issues/69
https://github.com/readysitego/spgo/issues/70

Sample app to reproduce:
```javascript

var sppull = require("sppull").sppull;

var context = {
  siteUrl: "https://tenant.sharepoint.com/search",
  creds: {
    username: "user@domain.com",
    password: "xxxx"
  }
};

var options = {
  camlCondition: '<And><Contains><FieldRef Name="FileDirRef"/><Value Type="Text">search/_catalogs/masterpage/display templates</Value></Contains><Contains><FieldRef Name="FileLeafRef"/><Value Type="Text">.html</Value></Contains></And>',
  createEmptyFolders:true,
  dlRootFolder:".",
  recursive:true,
  spBaseFolder:"/search",
  spDocLibUrl:"/_catalogs/masterpage/display templates"
};

sppull(context, options)
  .then(function(downloadResults) {
    console.log("Files are downloaded");
    console.log("For more, please check the results", JSON.stringify(downloadResults));
  })
  .catch(function(err) {
    console.log("Core error has happened", err);
  });
```

**Comment by koltyakov** [comment]
Will take a look, @sg-chrishasz

**Comment by koltyakov** [comment]
Looks like an error in the configuration:

```typescript
const options = {
  camlCondition: `<And><Contains><FieldRef Name="FileDirRef"/><Value Type="Text">search/_catalogs/masterpage/display templates</Value></Contains><Contains><FieldRef Name="FileLeafRef"/><Value Type="Text">.html</Value></Contains></And>`,
  // ...
  spBaseFolder: "/search",
  spDocLibUrl: "/_catalogs/masterpage/display templates" // <-- this is not a Document Library relative URL but a folder, should be Library for usage together with camlCondition
};
```

This should fail because `get list` by `/search/_catalogs/masterpage/display templates` will end up with incapability to distinguish the library. `spDocLibUrl: "/_catalogs/masterpage"` should be used instead.

Are you also sure that filtering based on `FileDirRef` is not blocked in the API? If I'm not wrong it's blocked at least in SPO and therefore 2019, probably 2016 too but not sure.

To make things simple we can add `fileRegExp` option with RegExp and use it with a combination of getting from a folder method like it is [done in another library](https://github.com/koltyakov/sppurge/blob/master/test/deleteUsingRegExp.ts#L14).

**Comment by sg-chrishasz** [comment]
It is not so much that the request is failing, because the files are all being downloaded. What I'm seeing is that when the query is executed against a subweb and the file path contains the SPWeb name, that specific name/string is being removed from the resulting file path or folder structure.

Thinking long-term about this. Would you rather support regexp for querying arbitrary files or CAML? The goal with this feature is that a user could download files based on a query pattern (e.g. `/siteassets/*.js`). Whether I use Glob or regexp for this is equally valuable from an end-user standpoint.

In the meantime, I'll work on some more test cases with sppull and see if I can find either a workaround or locate the source of the issue.

**Also**: Thanks for the notes on `FileDirRef` - I'm using that property, combined with the CAML query, in order to map glob notation to a SharePoint API query and I'll need to choose a different property if `FileDirRef` will be blocked.

**Comment by koltyakov** [comment]
I've added fileRegExp option. [Here](https://github.com/koltyakov/sppull/blob/master/test/manual/regexp.ts) is an example of usage.

```typescript
const pullOptions: ISPPullOptions = {
  spBaseFolder: '/',
  spRootFolder: '_catalogs/masterpage/display%20templates',
  dlRootFolder: join(__dirname, 'Downloads'),
  fileRegExp: new RegExp('(.*)/(.*)\.(xsl|xml)', 'i')
};
```

Library dependency update is required.

Please let me know if this is a kind of help with your scenario.

A possible downside is download time. As with reg expression and root folder option, the lib first iterates through the folders getting the list of files. Then filters the list based on regexp of any complexity.

A possible note is that the config is damn simple - folder paths and regexp, no need to struggle with CAML queries.

**Comment by sg-chrishasz** [comment]
Wow, great turn-around @koltyakov! 

I'm going to add regex support, in addition to current glob support, to SPGo and publish an update.

Thanks, as always, for the support!

**Comment by sg-chrishasz** [comment]
I've been working on RegEx support for SPGo tonight, and I'm not able to get any matches working. I'm sure it is user error, but I was wondering if you can help point me in the correct direction.

**Scenario (Digest Auth):**
SharePoint Url: `https://xxx.sharepoint.com/sites/spgo`
Download: all files and folders in the SiteAssets folder
Equivelent Glob: `/SiteAssets/**/*`

**SPPull config:**
Context
```json
{
    "creds" :  
    {
        "password": "xxxx", 
        "username": "username@domain.com"
    },
    "siteUrl" : "https://xxx.sharepoint.com/sites/spgo"
}
```

Options
```json
{
    "createEmptyFolders" : true,
    "dlRootFolder" : "c:\\Users\\username\\Documents\\projects\\test\\src",
    "fileRegExp" : "/^\/SiteAssets\/((?:[^\/]*(?:\/|$))*)([^\/]*)$/i {lastIndex: 0}",
    "recursive" : true,
    "spBaseFolder" : "/sites/spgo",
    "spRootFolder" : "/SiteAssets"
}

**Comment by sg-chrishasz** [comment]
oops, didn't mean to close.

**Comment by sg-chrishasz** [comment]
One more update on this ticket- I located the source of the original error and verified that the CAML query will work with a simple code change. 

After SPPull grabs the full list of files to download, there is a second pass in a function called `downloadFile` to format the `spFilePathRelative ` var. see here:
https://github.com/koltyakov/sppull/blob/4928444a1ffe045407fa2e771e4283eda9358721/src/api/index.ts#L43

This line uses a `RegExp` object to perform a `replace` function on the `spFilePath` var to trim the SP Web url stub from the `spFilePath` object. The replace is not limited to only checking the start of the string, so any occurrence of the SP Web url in the full string will be replaced. (e.g. if the SP Web URL is `/site1` and the file path is  /site1/SiteAssets/js/site1abc.xxx.js, the local directory url will be converted to:  `/SiteAssets/jsabc.xxx.js`). The replacement at the beginning of the string is correct, the second one is not needed.

To fix this, the Regex created on line 40 would need to specify that the replace should only be performed on a match at the beginning of the string, using the `^` token:
https://github.com/koltyakov/sppull/blob/4928444a1ffe045407fa2e771e4283eda9358721/src/api/index.ts#L40
---becomes---
```TypeScript
const spBaseFolderRegEx = new RegExp(decodeURIComponent('^' + this.options.spBaseFolder), 'gi');
```
and the files will be downloaded correctly.

**Comment by sg-chrishasz** [comment]
@koltyakov Per my testing, this is resolved in SPPull 2.3.3. Thank you!
